### ISSUE TITLE
Can't change attribute sets via csv import [Magento 2.1.0]

### ISSUE THREAD

**Issue by RG-1** [magento/magento2#5498]
## Steps to reproduce
1. Magento 2.1.0
2. Go to Stores > Attribute: Attribute Sets and create two new sets (e.g. S-1 and S-2)
3. Assign some attributes to sets
4. Create a csv file with products. Value of `attribute_set_code` should be "S-1"
5. Go to System > Data transfer: Import
6. Import products [Import Behavior: Add/Update]
7. Edit csv file. Change value of `attribute_set_code` to "S-2"
8. Go to System > Data transfer: Import
9. Import products [Import Behavior: Add/Update] 
10. Nothing is changed, so import products one more time, but change Import Behavior to Replace
## Expected result
1. Since Magento 2.1.0 it is possible to change attribute sets when editing products. I think that it should be possible on products update via import.
## Actual result
1. Attribute Sets are not changed.
2. Imports are successful, without any errors.
3. Only way to change attributes set is to delete products with old set and import products with new set.
## Example files

``` csv
"sku","store_view_code","attribute_set_code","product_type","categories","product_websites","name","weight","product_online","tax_class_name","visibility","price","url_key","display_product_options_in","msrp_display_actual_price_type","country_of_manufacture","qty","out_of_stock_qty","use_config_min_qty","is_qty_decimal","allow_backorders","use_config_backorders","min_cart_qty","use_config_min_sale_qty","max_cart_qty","use_config_max_sale_qty","is_in_stock","notify_on_stock_below","use_config_notify_stock_qty","manage_stock","use_config_manage_stock","use_config_qty_increments","qty_increments","use_config_enable_qty_inc","enable_qty_increments","is_decimal_divided","website_id","deferred_stock_update","use_config_deferred_stock_update"
"Simple1",,"S-1","simple","Default Category/issue",,"Simple 1",1,1,"Taxable Goods","Catalog, Search",10,,"Block after Info Column","Use config",,5,0,1,0,0,1,1,0,0,1,1,,1,0,1,1,0,1,0,0,1,0,1
"Simple2",,"S-1","simple","Default Category/issue",,"Simple 2",1,1,"Taxable Goods","Catalog, Search",20,,"Block after Info Column","Use config",,5,0,1,0,0,1,1,0,0,1,1,,1,0,1,1,0,1,0,0,1,0,1
"Simple3",,"S-1","simple","Default Category/issue",,"Simple 3",1,1,"Taxable Goods","Catalog, Search",30,,"Block after Info Column","Use config",,5,0,1,0,0,1,1,0,0,1,1,,1,0,1,1,0,1,0,0,1,0,1
```

``` csv
"sku","store_view_code","attribute_set_code","product_type","categories","product_websites","name","weight","product_online","tax_class_name","visibility","price","url_key","display_product_options_in","msrp_display_actual_price_type","country_of_manufacture","qty","out_of_stock_qty","use_config_min_qty","is_qty_decimal","allow_backorders","use_config_backorders","min_cart_qty","use_config_min_sale_qty","max_cart_qty","use_config_max_sale_qty","is_in_stock","notify_on_stock_below","use_config_notify_stock_qty","manage_stock","use_config_manage_stock","use_config_qty_increments","qty_increments","use_config_enable_qty_inc","enable_qty_increments","is_decimal_divided","website_id","deferred_stock_update","use_config_deferred_stock_update"
"Simple1",,"S-2","simple","Default Category/issue",,"Simple 1",1,1,"Taxable Goods","Catalog, Search",10,,"Block after Info Column","Use config",,5,0,1,0,0,1,1,0,0,1,1,,1,0,1,1,0,1,0,0,1,0,1
"Simple2",,"S-2","simple","Default Category/issue",,"Simple 2",1,1,"Taxable Goods","Catalog, Search",20,,"Block after Info Column","Use config",,5,0,1,0,0,1,1,0,0,1,1,,1,0,1,1,0,1,0,0,1,0,1
"Simple3",,"S-2","simple","Default Category/issue",,"Simple 3",1,1,"Taxable Goods","Catalog, Search",30,,"Block after Info Column","Use config",,5,0,1,0,0,1,1,0,0,1,1,,1,0,1,1,0,1,0,0,1,0,1
```

**Comment by BaDos** [comment]
Hi @RG-1 

Thank you for reporting.
The internal ticket MAGETWO-58385 was created to resolve this issue.

**Comment by isitnikov** [comment]
@RG-1  please try such workaround:
```
diff --git a/vendor/magento/module-catalog-import-export/Model/Import/Product.php b/vendor/magento/module-catalog-import-export/Model/Import/Product.php
index 970b8a1..f6e7961 100644
--- a/vendor/magento/module-catalog-import-export/Model/Import/Product.php
+++ b/vendor/magento/module-catalog-import-export/Model/Import/Product.php
@@ -1296,7 +1296,7 @@ class Product extends \Magento\ImportExport\Model\Import\Entity\AbstractEntity
             $entityTable = $this->_resourceFactory->create()->getEntityTable();
         }
         if ($entityRowsUp) {
-            $this->_connection->insertOnDuplicate($entityTable, $entityRowsUp, ['updated_at']);
+            $this->_connection->insertOnDuplicate($entityTable, $entityRowsUp, ['updated_at', 'attribute_set_id']);
         }
         if ($entityRowsIn) {
             $this->_connection->insertMultiple($entityTable, $entityRowsIn);
@@ -1478,7 +1478,12 @@ class Product extends \Magento\ImportExport\Model\Import\Entity\AbstractEntity
                         'updated_at' => (new \DateTime())->format(DateTime::DATETIME_PHP_FORMAT),
                         $this->getProductEntityLinkField()
                             => $this->_oldSku[$rowSku][$this->getProductEntityLinkField()],
+                        'attribute_set_id' => isset($this->_attrSetNameToId[$rowData['attribute_set_code']])
+                            ? $this->_attrSetNameToId[$rowData['attribute_set_code']]
+                            : $this->_oldSku[$rowSku]['attr_set_id'],
                     ];
+
+                    $rowData[self::COL_ATTR_SET] = $rowData['attribute_set_code'];
                 } else {
                     if (!$productLimit || $productsQty < $productLimit) {
                         $entityRowsIn[$rowSku] = [
```

**Comment by antbates91** [comment]
@isakov I've noticed that the following patch will error if a csv is imported without the `attribute_set_code` column.

Other than that it fixes the issue and you can import changes to Attribute Set Codes without issue.

**Comment by jameswithers** [comment]
This is still an issue in Community 2.1.6: https://github.com/magento/magento2/issues/10167.

The patch provided works.

**Comment by magento-engcom-team** [comment]
@RG-1, thank you for your report.
The issue is already fixed in 2.2.0
